File: 백준/실버/2667. 단지번호붙이기/단지번호붙이기.cpp
#include <iostream>
#include <algorithm>
using namespace std;

int dx[4] = {0, 0, -1, 1};
int dy[4] = {1, -1, 0, 0};
int visited[26][26];
int map[26][26];
int n;
int cnt;
vector<int> cnts;

void dfs(int x, int y){
    visited[x][y] = 1;
    cnt++;
    for(int i = 0; i < 4; ++i){
        int nx = x + dx[i];
        int ny = y + dy[i];
        if(visited[nx][ny]) continue;
        if(nx < 1 || nx > n || ny > n || ny < 1 || map[nx][ny] != 1) continue;
        dfs(nx, ny);
    }
}

int main(){
    ios_base::sync_with_stdio(false); cin.tie(NULL); cout.tie(NULL);
    cin >> n;
    for(int i = 1; i <= n; ++i){
        for(int j = 1; j <= n; ++j){
            char c;
            cin >> c;
            map[i][j] = (c == '1');
        }
    }

    for(int i = 1; i <= n; ++i){
        for(int j = 1; j <= n; ++j){
            if(!visited[i][j] && map[i][j] == 1){
                dfs(i, j);
                cnts.push_back(cnt);
                cnt = 0;
            }
        }
    }
    cout << cnts.size() << '\n';
    sort(cnts.begin(), cnts.end());
    for(int i = 0; i < cnts.size(); ++i){
        cout << cnts[i] << '\n';
    }
}

File: 백준/실버/2667. 단지번호붙이기/단지번호붙이기.js
const fs = require("fs");
const stdin = fs.readFileSync("ex.txt").toString().trim().split("\n");

const N = +stdin[0];
const arr = stdin.slice(1).map((str) => str.split("").map(Number));

const dx = [0, 0, -1, 1];
const dy = [1, -1, 0, 0];

function solution(N, arr) {
  const visited = Array.from({ length: N + 1 }, () => Array(N + 1).fill(0));
  const answer = [];

  function DFS(x, y) {
    let cnt = 1;
    for (let i = 0; i < 4; ++i) {
      const nx = x + dx[i];
      const ny = y + dy[i];

      if (nx < 0 || nx >= N || ny < 0 || ny >= N || arr[nx][ny] === 0) continue;
      if (visited[nx][ny]) continue;

      visited[nx][ny] = 1;
      cnt += DFS(nx, ny);
    }
    return cnt;
  }

  let cnt = 0;
  for (let i = 0; i < N; ++i) {
    for (let j = 0; j < N; ++j) {
      if (!visited[i][j] && arr[i][j] !== 0) {
        visited[i][j] = 1;
        const home_cnt = DFS(i, j);
        cnt++;
        answer.push(home_cnt);
      }
    }
  }
  answer.sort((a, b) => a - b);

  [cnt].concat(answer).forEach((num) => console.log(num));
}

solution(N, arr);
